### ISSUE TITLE
Very slow startup time while getting device pixel ratio

### ISSUE THREAD

**Issue by Hubro** [alacritty/alacritty#5608]
I have just set up Linux on a new laptop and everything seems to be working fine, with one exception. Alacritty is spending a shocking amount of time starting up a new window.

If I launch a new Alacritty window from a different terminal emulator (e.g. xcfe4-terminal) or `bemenu-run`, it can take 3-5 seconds to launch the window. However, if I launch Alacritty from an existing Alacritty window, it will start more or less instantly.

### System

OS: Arch Linux
Version: `alacritty 0.9.0 (fed349aa)`
Linux/BSD: Wayland, sway version 1.6.1

### Logs

Launching Alacritty from xfce4-terminal:

```
$ alacritty -vvv -e false
Created log file at "/tmp/Alacritty-384178.log"
[2021-11-14 12:51:40.082743596] [INFO ] [alacritty] Welcome to Alacritty
[2021-11-14 12:51:40.082785509] [INFO ] [alacritty] Configuration files loaded from:
                                                      "/home/tomas/.config/alacritty/alacritty.yml"
[2021-11-14 12:51:40.097857501] [TRACE] [crossfont] Number of fonts is 183
[2021-11-14 12:51:40.097976900] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-Regular.otf", index=0
[2021-11-14 12:51:40.098129349] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:51:40.100298347] [DEBUG] [alacritty] Estimated DPR: 1
[2021-11-14 12:51:40.100311700] [DEBUG] [alacritty] Estimated window size: None
[2021-11-14 12:51:40.100320033] [DEBUG] [alacritty] Estimated cell size: 7 x 16
[2021-11-14 12:51:44.031082609] [INFO ] [alacritty] Device pixel ratio: 1
[2021-11-14 12:51:44.036104549] [INFO ] [alacritty] Initializing glyph cache...
[2021-11-14 12:51:44.039913255] [TRACE] [crossfont] Number of fonts is 183
[2021-11-14 12:51:44.039962861] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-Regular.otf", index=0
[2021-11-14 12:51:44.040095390] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:51:44.044657609] [TRACE] [crossfont] Number of fonts is 181
[2021-11-14 12:51:44.044706630] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-Bold.otf", index=0
[2021-11-14 12:51:44.044786285] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:51:44.049158772] [TRACE] [crossfont] Number of fonts is 185
[2021-11-14 12:51:44.049204371] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-It.otf", index=0
[2021-11-14 12:51:44.049321132] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:51:44.053586362] [TRACE] [crossfont] Number of fonts is 186
[2021-11-14 12:51:44.053626974] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-BoldIt.otf", index=0
[2021-11-14 12:51:44.053725387] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:51:44.066405739] [INFO ] [alacritty] ... finished initializing glyph cache in 0.030282126s
[2021-11-14 12:51:44.066433491] [INFO ] [alacritty] Cell size: 7 x 16
[2021-11-14 12:51:44.066440460] [INFO ] [alacritty] Padding: 0 x 0
[2021-11-14 12:51:44.066469205] [INFO ] [alacritty] Width: 800, Height: 600
[2021-11-14 12:51:44.067471787] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 60, g: 56, b: 54 } to Rgb { r: 39, g: 36, b: 35 }
[2021-11-14 12:51:44.067488596] [TRACE] [alacritty] Deriving dim colors from normal colors
[2021-11-14 12:51:44.067494629] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 251, g: 241, b: 199 } to Rgb { r: 165, g: 159, b: 131 }
[2021-11-14 12:51:44.067524313] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 204, g: 36, b: 29 } to Rgb { r: 134, g: 23, b: 19 }
[2021-11-14 12:51:44.067530498] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 152, g: 151, b: 26 } to Rgb { r: 100, g: 99, b: 17 }
[2021-11-14 12:51:44.067536530] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 215, g: 153, b: 33 } to Rgb { r: 141, g: 100, b: 21 }
[2021-11-14 12:51:44.067542193] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 69, g: 133, b: 136 } to Rgb { r: 45, g: 87, b: 89 }
[2021-11-14 12:51:44.067548123] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 177, g: 98, b: 134 } to Rgb { r: 116, g: 64, b: 88 }
[2021-11-14 12:51:44.067553976] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 104, g: 157, b: 106 } to Rgb { r: 68, g: 103, b: 69 }
[2021-11-14 12:51:44.067563657] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 124, g: 111, b: 100 } to Rgb { r: 81, g: 73, b: 66 }
[2021-11-14 12:51:44.067603075] [INFO ] [alacritty] PTY dimensions: 37 x 114
[2021-11-14 12:51:44.070196074] [INFO ] [alacritty] Initialisation complete
[2021-11-14 12:51:44.074930279] [DEBUG] [alacritty_terminal] New num_cols is 132 and num_lines is 63
[2021-11-14 12:51:44.078234967] [INFO ] [alacritty] Padding: 0 x 0
[2021-11-14 12:51:44.078274050] [INFO ] [alacritty] Width: 928, Height: 1010
[2021-11-14 12:51:44.099588968] [INFO ] [alacritty] Goodbye
Deleted log file at "/tmp/Alacritty-384178.log"
```

Notice 99.9% of the start-up time is spent right before "Device pixel ratio: 1" is logged.

Alacritty run from an existing Alacritty window:

```
➜ alacritty -vvv -e false
Created log file at "/tmp/Alacritty-394758.log"
[2021-11-14 12:55:22.048846896] [INFO ] [alacritty] Welcome to Alacritty
[2021-11-14 12:55:22.048891559] [INFO ] [alacritty] Configuration files loaded from:
                                                      "/home/tomas/.config/alacritty/alacritty.yml"
[2021-11-14 12:55:22.063719250] [TRACE] [crossfont] Number of fonts is 183
[2021-11-14 12:55:22.063758101] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-Regular.otf", index=0
[2021-11-14 12:55:22.063864055] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:55:22.066029021] [DEBUG] [alacritty] Estimated DPR: 1
[2021-11-14 12:55:22.066042789] [DEBUG] [alacritty] Estimated window size: None
[2021-11-14 12:55:22.066050898] [DEBUG] [alacritty] Estimated cell size: 7 x 16
[2021-11-14 12:55:22.138792895] [INFO ] [alacritty] Device pixel ratio: 1
[2021-11-14 12:55:22.145011650] [INFO ] [alacritty] Initializing glyph cache...
[2021-11-14 12:55:22.149013198] [TRACE] [crossfont] Number of fonts is 183
[2021-11-14 12:55:22.149045772] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-Regular.otf", index=0
[2021-11-14 12:55:22.149124636] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:55:22.154077032] [TRACE] [crossfont] Number of fonts is 181
[2021-11-14 12:55:22.154124072] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-Bold.otf", index=0
[2021-11-14 12:55:22.154275767] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:55:22.158474456] [TRACE] [crossfont] Number of fonts is 185
[2021-11-14 12:55:22.158514576] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-It.otf", index=0
[2021-11-14 12:55:22.158655078] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:55:22.163163334] [TRACE] [crossfont] Number of fonts is 186
[2021-11-14 12:55:22.163203876] [TRACE] [crossfont] Got font path="/usr/share/fonts/adobe-source-code-pro/SourceCodePro-BoldIt.otf", index=0
[2021-11-14 12:55:22.163318616] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-11-14 12:55:22.175622613] [INFO ] [alacritty] ... finished initializing glyph cache in 0.03058819s
[2021-11-14 12:55:22.175649593] [INFO ] [alacritty] Cell size: 7 x 16
[2021-11-14 12:55:22.175657356] [INFO ] [alacritty] Padding: 0 x 0
[2021-11-14 12:55:22.175687390] [INFO ] [alacritty] Width: 800, Height: 600
[2021-11-14 12:55:22.177047483] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 60, g: 56, b: 54 } to Rgb { r: 39, g: 36, b: 35 }
[2021-11-14 12:55:22.177067413] [TRACE] [alacritty] Deriving dim colors from normal colors
[2021-11-14 12:55:22.177074421] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 251, g: 241, b: 199 } to Rgb { r: 165, g: 159, b: 131 }
[2021-11-14 12:55:22.177104866] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 204, g: 36, b: 29 } to Rgb { r: 134, g: 23, b: 19 }
[2021-11-14 12:55:22.177111516] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 152, g: 151, b: 26 } to Rgb { r: 100, g: 99, b: 17 }
[2021-11-14 12:55:22.177117936] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 215, g: 153, b: 33 } to Rgb { r: 141, g: 100, b: 21 }
[2021-11-14 12:55:22.177124109] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 69, g: 133, b: 136 } to Rgb { r: 45, g: 87, b: 89 }
[2021-11-14 12:55:22.177130674] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 177, g: 98, b: 134 } to Rgb { r: 116, g: 64, b: 88 }
[2021-11-14 12:55:22.177136893] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 104, g: 157, b: 106 } to Rgb { r: 68, g: 103, b: 69 }
[2021-11-14 12:55:22.177147050] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 124, g: 111, b: 100 } to Rgb { r: 81, g: 73, b: 66 }
[2021-11-14 12:55:22.177188441] [INFO ] [alacritty] PTY dimensions: 37 x 114
[2021-11-14 12:55:22.179611242] [INFO ] [alacritty] Initialisation complete
[2021-11-14 12:55:22.184311265] [DEBUG] [alacritty_terminal] New num_cols is 132 and num_lines is 63
[2021-11-14 12:55:22.187388986] [INFO ] [alacritty] Padding: 0 x 0
[2021-11-14 12:55:22.187439299] [INFO ] [alacritty] Width: 928, Height: 1010
[2021-11-14 12:55:22.206342493] [INFO ] [alacritty] Goodbye
Deleted log file at "/tmp/Alacritty-394758.log"
```

**Comment by yujiri8** [comment]
I have similar symptoms:
* Alacritty takes 0.5-1 second to launch, but for me it happens regardless of where it's launched from
* The delay happens between "Estimated cell size" and "Device pixel ratio"

OS: Artix Linux
Alacritty: 0.9.0
Wayland, sway 1.6.1

Alacritty used to start instantly for me, but after an update, I think it was a couple weeks ago, this started happening. Other terminals such as xterm start instantly in the same environment.

**Comment by chrisduerr** [comment]
> if I launch Alacritty from an existing Alacritty window, it will start more or less instantly.

What exactly do you mean by that? Version 0.9.0 didn't have multiwindow yet so I don't see how that would be different?

Sounds like the issue might be related to the environment, so comparing that might be interesting. Are you launching both instances as Wayland native, or are you running one through XWayland? If not, does running it through XWayland fix this?

> Alacritty used to start instantly for me, but after an update, I think it was a couple weeks ago, this started happening. Other terminals such as xterm start instantly in the same environment.

If something happens after an update, it would be nice to know *what* you updated.

**Comment by Hubro** [comment]
> > if I launch Alacritty from an existing Alacritty window, it will start more or less instantly.
> 
> What exactly do you mean by that? Version 0.9.0 didn't have multiwindow yet so I don't see how that would be different?

I did this dozens of time yesterday, and consistently Alacritty would start quickly when launched from another instance of Alacritty. However today I can no longer replicate this, now it takes several seconds to launch no matter what.

Although, if I launch Alacritty several times in quick succession, only the first launch will be super slow, so there seems to be some caching going on somewhere.

Is there any way to increase verbosity even further, so we can see what's taking so long?

Since this is only happening on this one system (all my other PCs start Alacritty just fine) I agree it's probably environmental. Do you know how Alacritty calculates the device pixel ratio? Is there any way I can override it to skip whatever Alacritty does to calculate it?

**Comment by yujiri8** [comment]
> If something happens after an update, it would be nice to know what you updated.

Sorry, I don't actually know what updated because it was a full system upgade `pacman -Syu`. So alacritty probably got updated in there, but I can't know what version it was prior.

**Comment by adroslice** [comment]
> > If something happens after an update, it would be nice to know what you updated.
> 
> Sorry, I don't actually know what updated because it was a full system upgade `pacman -Syu`. So alacritty probably got updated in there, but I can't know what version it was prior.

There are logs for that: `/var/log/pacman.log`
I have the same problem, though my issue was closed. Downgrading alacritty didn't help though.
Definitely environmental.

**Comment by Hubro** [comment]
Is there anything I can do to troubleshoot this further? The 4-5 second startup time is killing me. I didn't realize how often I open terminals before now.

**Comment by adroslice** [comment]
I think I just figured it out: Downgrading `egl-wayland` from `1:1.1.9+gdaab854-1` to `1:1.1.7-1` resolved the issue for me. Please check your /var/cache/pacman/pkg for your old version.

**Comment by Hubro** [comment]
> I think I just figured it out: Downgrading `egl-wayland` (required by nvidia-utils <- wlroots <- sway) from `1:1.1.9+gdaab854-1` to `1:1.1.7-1` resolved the issue for me. Please check your `/var/cache/pacman/pkg/` for your old version.

Yes! You're a f***ing hero!

I downloaded and installed 1.1.7 from [the archive](https://archive.archlinux.org/packages/e/egl-wayland/) and it *immediately* fixed the problem. I also tried 1.1.8 and the problem returned, so whatever is causing this, it was introduced in 1.1.8.

**Comment by chrisduerr** [comment]
> I think I just figured it out: Downgrading egl-wayland (required by nvidia-utils <- wlroots <- sway) from 1:1.1.9+gdaab854-1 to 1:1.1.7-1 resolved the issue for me. Please check your /var/cache/pacman/pkg/ for your old version.

Yeah, nvidia on sway so not much of a surprise there.

**Comment by vberger** [comment]
FYI, there was an issue in SCTK's code causing at least some of the slowdowns. I have now released SCTK version 0.15.3 fixing that problem.

**Comment by chrisduerr** [comment]
Thanks for the heads-up.
